bokehjs/src/coffee/models/plots/plot_canvas.coffee

```diff
@@ -402,7 +402,12 @@ class PlotCanvasView extends Renderer.View
     # reach the frame in time (sometimes) so force an update here for now
     @model.get('frame')._update_mappers()
 
+    # Get context and restore to its initial state. The restoring
+    # Would stricly speaking not be necessary, but in case there is an error
+    # during drawing, restore() won't be called enough. This guards for that.
     ctx = @canvas_view.ctx
+    for i in [0..3]
+      ctx.restore()
```

Comment:
Right. Now I understand your concern. However, I think there has always been an implicit assumption that `plot.render()` was the "rendering entry point" and that there were no calls to `save()` before it, because `plot.render()` calls `@canvas_view.prepare_canvas()` (previously `@view_canvas.render()`) which may resize the canvas. And when a canvas is resized, the state of a context is reset. So we probably should add a comment about this assumption.

An alternative might be to always set the size of the canvas in `@canvas_view.prepare_canvas()` but I read somewhere that this might be unreliable in some browsers if the size is not changed. This silly for-loop is my version of `ctx.reset()`. I agree it does not look pretty, but it seemed the easiest way to get the job done, without having to change any other code.

Using `try` / `catch` could work, though I don't know what that would do to tracebacks, since there is no `finally` (right?).